Improvements

drakkan opened the pull request:
The commit messages should be self explanatory,

please let me know what do you think about, thanks!

drakkan commented:
for the unlinked file support on windows we could use this library

https://github.com/calebcase/tmpfile

If you prefer this way I'll send a new pull request, thanks!

eikenb commented:
Hey @drakkan! Glad you found this little library useful. Sorry for the delay, been laid up with a bad cold and not checking my email.

Thanks for putting the linter fixes in their own commit. Another instance of my snake_case habits from my python days coming back.

For the PipeInDir commit, I'd appreciate a simple test that exercises that code. Really just needs to make the call and make sure it gets a valid FD/doesn't get an error.

For windows compatibility I'm not sure. Removing the file when done increases the chance for leaving around garbage files but using the calebcase/tmpfile package would introduce the first external dependency and I liked not having any (not counting testing). Is your preference just to remove it as you've done or to use calebcase/tmpfile?

Probably a lame idea, but would opening/closing/reopening (with proper flags) be to much overhead? Using the built-in ioutil.Tempfile() as a way to generate the file and then re-opening it with the proper flags.

drakkan commented:
> Hey @drakkan! Glad you found this little library useful. Sorry for the delay, been laid up with a bad cold and not checking my email.
> 
> Thanks for putting the linter fixes in their own commit. Another instance of my snake_case habits from my python days coming back.
> 
> For the PipeInDir commit, I'd appreciate a simple test that exercises that code. Really just needs to make the call and make sure it gets a valid FD/doesn't get an error.

Well this is already tested, now Pipe() calls PipeInDir with an empty directory

https://github.com/eikenb/pipeat/pull/1/commits/e013e687692b78902dab3c46526c7f4b41ae31b1#diff-827ca6cdff8da68d0e00a2c65ee6417fR114

test coverage is unchanged, eventually a negative test could be added to test that if the specified directory does not exists PipeInDir will return an error

> 
> For windows compatibility I'm not sure. Removing the file when done increases the chance for leaving around garbage files but using the calebcase/tmpfile package would introduce the first external dependency and I liked not having any (not counting testing). Is your preference just to remove it as you've done or to use calebcase/tmpfile?
> 
> Probably a lame idea, but would opening/closing/reopening (with proper flags) be to much overhead? Using the built-in ioutil.Tempfile() as a way to generate the file and then re-opening it with the proper flags.

I'll try this thanks for the suggestion.

I have a question/concern too:

- suppose you have to upload a big file, the SFTP server is on your local LAN and you have a slow connection to internet and so to AWS, in this case you have to wait for the transfer to complete before replying to the client. So you have to use a sync Pipe, the SFTP client will quickly send all the data and will remain a lot of time blocked waiting for the remote transfer to complete. I suspect that this could cause issues for some SFTP client, they could timeout. Do you have tested this?
I'm thinking to add a new option that allows to set the maximum bytes difference between the writer and the reader and block the writer is the reader is too slow, what do you think about? thanks

eikenb commented:
Regarding your question... yes, that is an issue. The design was targeting servers running in the cloud with lots of bandwidth to their data point. So that the bottleneck would be between the client and the server, not the server and the storage.

A max byte difference would be one way to tackle that. If you have something already in mind give it a go. I haven't worked with this library in a while so you probably have a better mental model of it at the moment than I and if that sounds like the best path to you...

drakkan commented:
@eikenb I implemented the file unlink, on windows, as you suggested and I added a check for unlink errors too.
Before the test cases worked on Windows, since the os.Remove error was ignored, but the temporary file was not removed, should be fixed now.

I added basic test case for PipeInDir/AsyncWriterPipeInDir too.
Please let me know what is missing, thanks!

drakkan commented:
@eikenb I think I have fixed the speed issue too

If the pipe is not async, the writer cannot write more than the reader requested, I added a write condition to signal the writers from the reader.

Readers and writers are now syncronized on `dataLock`. I added some test cases to cover the new code.

I did several tests, even on a windows virtual machine, and all seems fine. I'm using this branch in SFTP Go now, let's see if some users complaint

drakkan commented:
a performance improvement here too!

eikenb commented:
@drakkan Sorry... I'll try to take a look. Life continues to be pretty crazy, but I'm trying to carve out a little time each week to make some progress so please be patient.

Thanks for your patients and thanks again for the contribution. I do appreciate both.

drakkan commented:
@eikenb no problem, I can continue to use my branch until you find some time to review my changes.

eikenb commented:
Everything looks great so far. I want to spend a bit more time reading over the changes in the "the writer now blocks if it is quicker than the reader" commit, but my initial impression is that they look good. I'm going to pull them down and play with it some to be sure I understand the changes.

Thanks.

eikenb commented on pipeat.go:
```diff
@@ -32,34 +32,65 @@ func (c spans) Less(i, j int) bool { return c[i].start < c[j].start }
+		select {
+		case <-f.eow:
+			trace("eow")
+			break L
```
Style preference.... I prefer to use returns when possible instead of breaks. Could you change the use of breaks like this to returns?

eikenb commented:
Finally got a break to read through this enough to understand what's going on. The tests really helped. Anyways, I'm good with this and am merging. Thanks!
